README.md
# lush
Shell project for University Task (luciano shell = lush)


functions/history.txt
clear
clear
ls
clear
history
history
ls
history
clear
history
ls
history
clear


functions/loop.h
#ifndef LOOP_H
#define LOOP_H

#include "funcs.h"
#include <limits.h>
#include "lib/helper.h"

int lush_launch(char **args)
{
  pid_t pid;
  int status;

  pid = fork();
  if (pid == 0) {
    // Child process
    if (execvp(args[0], args) == -1) {
      perror("lsh");
    }
    exit(EXIT_FAILURE);
  } else if (pid < 0) {
    // Error forking
    perror("lsh");
  } else {
    // Parent process
    do {
      waitpid(pid, &status, WUNTRACED);
    } while (!WIFEXITED(status) && !WIFSIGNALED(status));
  }

  return 1;
}

int lush_execute(char **args)
{
  int i;
  FILE* history;
  char buffer[100];
  int cx;
  const char* s = getenv("HOME");
  cx = snprintf ( buffer, 100, "%s/history.txt", s);
  history = fopen(buffer, "a");

  if (args[0] == NULL) {
    return 1;
  }

  for (i = 0; i < lsh_num_builtins(); i++) {
    if (strcmp(args[0], builtin_str[i]) == 0) {
      fputs(builtin_str[i], history);
      fputs("\n", history);
      fclose(history);
      return (*builtin_func[i])(args);
    }
  }

  return lush_launch(args);
}

#define LSH_RL_BUFSIZE 1024

char *lsh_read_line(void)
{
  int bufsize = LSH_RL_BUFSIZE;
  int position = 0;
  char *buffer = malloc(sizeof(char) * bufsize);
  int c;

  if (!buffer) {
    fprintf(stderr, "lsh: allocation error\n");
    exit(EXIT_FAILURE);
  }

  while (1) {
    // Read a character
    c = getchar();

    if (c == EOF) {
      exit(EXIT_SUCCESS);
    } else if (c == '\n') {
      buffer[position] = '\0';
      return buffer;
    } else {
      buffer[position] = c;
    }
    position++;

    // If we have exceeded the buffer, reallocate.
    if (position >= bufsize) {
      bufsize += LSH_RL_BUFSIZE;
      buffer = realloc(buffer, bufsize);
      if (!buffer) {
        fprintf(stderr, "lsh: allocation error\n");
        exit(EXIT_FAILURE);
      }
    }
  }
}

#define LSH_TOK_BUFSIZE 64
#define LSH_TOK_DELIM " \t\r\n\a"

char **lsh_split_line(char *line)
{
  int bufsize = LSH_TOK_BUFSIZE, position = 0;
  char **tokens = malloc(bufsize * sizeof(char*));
  char *token, **tokens_backup;

  if (!tokens) {
    fprintf(stderr, "lsh: allocation error\n");
    exit(EXIT_FAILURE);
  }

  token = strtok(line, LSH_TOK_DELIM);
  while (token != NULL) {
    tokens[position] = token;
    position++;

    if (position >= bufsize) {
      bufsize += LSH_TOK_BUFSIZE;
      tokens_backup = tokens;
      tokens = realloc(tokens, bufsize * sizeof(char*));
      if (!tokens) {
		free(tokens_backup);
        fprintf(stderr, "lsh: allocation error\n");
        exit(EXIT_FAILURE);
      }
    }

    token = strtok(NULL, LSH_TOK_DELIM);
  }
  tokens[position] = NULL;
  return tokens;
}

/**
   @brief Loop getting input and executing it.
 */
void lsh_loop(void)
{
  char *line;
  char **args;
  int status;
  char *user=getenv("USER");
  if(!user){
    user = "default";
  }
  char hostname[1024];
  gethostname(hostname, 1024);

  do {
    char cwd[PATH_MAX];
    if (getcwd(cwd, sizeof(cwd)) != NULL) 
    {
      if(checkIfGit(cwd) == true){
        char *branch = getGitBranch(cwd);
        printf("\033[22;35m%s@%s \033[22;31m%s \033[22;36m(%s)\033[0m\n", user, hostname,cwd, branch);
      }
      else {
        printf("\033[22;35m%s@%s \033[22;31m%s\033[0m\n", user, hostname,cwd);
      }
    } 
    else 
    {
      perror("getcwd() error");

    }    
    printf("-> ");
    line = lsh_read_line();
    args = lsh_split_line(line);
    
    status = lush_execute(args);
    printf("\n");

    free(line);
    free(args);
  } while (status);
}


#endif 

functions/lib/helper.h
#ifndef HELPER_H
#define HELPER_H

#include <string.h>
#include <stdio.h>
#include <stdbool.h>

char* replaceSubString(char* string, char* oldSub, char* newSub){

  static char buffer[4096];
  char *p;

  if(!(p = strstr(string, oldSub)))  
    return string;

  strncpy(buffer, string, p-string); 
  buffer[p-string] = '\0';

  sprintf(buffer+(p-string), "%s%s", newSub, p+strlen(oldSub));

  return buffer;

}

bool checkIfGit(char* cwd){
  DIR *d;
  struct dirent *dir;
  d = opendir(cwd);
  while ((dir = readdir(d)) != NULL) {
      if(strcmp(dir->d_name,".git") == 0){
        return true;
      }
  }
  return false;
}

char* getGitBranch(char* path){
  FILE* fp;
  fp = fopen(".git/HEAD", "r");
  char *fileText = malloc(sizeof(char)* 50);


  fgets(fileText,100,fp);
  fclose(fp);
  strtok(fileText, "\n");
  strtok(fileText, "/");
  strtok(NULL, "/");
  
  return strtok(NULL, "/");
}

int countWords(char* string){

	int len,i,word=1;
	len=strlen(string);
	for(i=0;i<len;i++)
	{
		if(string[i]!=' ' && string[i+1]==' ')
			word=word+1;
	}

  return word;
}

#endif

functions/funcs.h
#ifndef FUNCS_H
#define FUNCS_H

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <dirent.h> 
#include <string.h>
#include <limits.h>
#include <pwd.h>
#include "lib/helper.h"


int lush_cd(char **args);
int lush_help(char **args);
int lush_exit(char **args);
int lush_ls(char **args);
int lush_clear(char **args);
int lush_history(char **args);
int lush_clhist(char **args);
int lush_pwd();
int lush_mkdir(char **args);
int lush_rmdir(char **args);
int lush_rm(char **args);
int lush_touch(char **args);
int lush_cat(char **args);


char *builtin_str[] = {
  "cd",
  "help",
  "exit",
  "ls",
  "clear",
  "history",
  "clhist",
  "pwd",
  "mkdir",
  "touch",
  "rm",
  "rmdir",
  "cat",
};

int (*builtin_func[]) (char **) = {
  &lush_cd,
  &lush_help,
  &lush_exit,
  &lush_ls,
  &lush_clear,
  &lush_history,
  &lush_clhist,
  &lush_pwd,
  &lush_mkdir,
  &lush_touch,
  &lush_rm,
  &lush_rmdir,
  &lush_cat
};

int lsh_num_builtins() {
  return sizeof(builtin_str) / sizeof(char *);
}

int lush_rm(char **args){
  int result = remove(args[1]);
  if(result != -1){
    return 1;
  }
  printf("Couldn't delete file");
  return 1;
}

int lush_rmdir(char **args){
  int result = 0;
  if(args[1]){
    result = rmdir(args[1]);
    if(result == -1){
      printf("Directory is not empty man");
    }
  }else{
    printf("Cant delete");  
  }
  return 1;
}

int lush_touch(char **args){
  FILE* newFile;
  newFile = fopen(args[1], "w");
  fclose(newFile);
  return 1;
}

int lush_mkdir(char **args){
  int result = 0;
  if(args[1]){
    result = mkdir(args[1], 0777);
  }
  return 1;
}

int lush_clhist(char **args){
  FILE *history;
  char buffer[100];
  int cx;
  const char* s = getenv("HOME");
  cx = snprintf ( buffer, 100, "%s/history.txt", s);
  history = fopen(buffer, "w");
  fputs("", history);
  fclose(history);
  return 1;
}

int lush_pwd(char **args){
  char cwd[PATH_MAX];
    if (getcwd(cwd, sizeof(cwd)) != NULL) 
    {
      printf("%s", cwd);
    } 
    else {
      printf("Error");
    }

    return 1;
  
}

int lush_history(char **args){
  char* status;
  char input[255];
  int counter = 1;
  FILE *history;
  char buffer[100];
  int cx;
  const char* s = getenv("HOME");
  cx = snprintf ( buffer, 100, "%s/history.txt", s);
  history = fopen(buffer, "r");
  
  do 
  {
    status = fgets(input, 255, history);
    if(status != 0){
      printf("%d. %s", counter, input);
    }
    counter++;
  } while(status);

  fclose(history);
  return 1;

}

int lush_ls(char **args){
  DIR *d;
  struct dirent *dir;
  if(args[1]){
    d = opendir(args[1]);
  }
  else{
    d = opendir(".");
  }
  if (d) {
    while ((dir = readdir(d)) != NULL) {
      if(strcmp(dir->d_name,".") != 0 && strcmp(dir->d_name, "..") != 0 ){
        if(countWords(dir->d_name) == 1){
          printf("\033[22;32m%s\033[0m ", dir->d_name);
        }
        else {
          printf("'\033[22;32m%s\033[0m' ", dir->d_name);
        }
      }
    }
    closedir(d);
  }
  printf("\n");
  return(1);
}

int lush_clear(char **args){
  system("@cls||clear");
  return(1);
}


int lush_cd(char **args)
{
  if (args[1] == NULL) {
    fprintf(stderr, "lush: expected argument to \"cd\"\n");
  }
  else if(strstr(args[1], "~") != NULL){
    char* s = getenv("HOME");
    char* result = replaceSubString(args[1],"~", s);
    if (chdir(result) != 0) {
      perror("lush");
    }
  }
  else {
    if (chdir(args[1]) != 0) {
      perror("lush");
    }
  }
  return 1;
}

int lush_help(char **args)
{
  int i;
  printf("Luciano Peranni's LUSH\n");
  printf("Type program names and arguments, and hit enter.\n");
  printf("The following are built in:\n");

  for (i = 0; i < lsh_num_builtins(); i++) {
    printf("  %s\n", builtin_str[i]);
  }

  printf("Use the man command for information on other programs.\n");
  return 1;
}

int lush_cat(char **args){
    FILE * fp;
    char * line = NULL;
    size_t len = 0;
    ssize_t read;

    
    fp = fopen(args[1], "r");
    if (fp == NULL)
        return 1;

    while ((read = getline(&line, &len, fp)) != -1) {
        printf("%s", line);
    }

    fclose(fp);
    if (line)
        free(line);
    return 1;
}

int lush_exit(char **args)
{
  return 0;
}


#endif